building_sim_plugins/building_plugins_common/include/building_sim_common/crowd_simulator_common.hpp

```diff
@@ -0,0 +1,208 @@
+#include <MengeCore/Orca/ORCASimulator.h>
+#include <MengeCore/PluginEngine/CorePluginEngine.h>
+
+namespace crowd_simulator {
+
+using AgentPtr = std::shared_ptr<Menge::Agents::BaseAgent>;
+//===============================================================
+/*
+* class AgentPose3d
+* replace the common interface of ignition::math::Pose3d, which might apply with different version in different plugins
+*/
+class  AgentPose3d{
+
+public:
+  AgentPose3d();
+  AgentPose3d(double& x, double& y, double& z, double& pitch, double& roll, double& yaw);
+
+  ~AgentPose3d();
+
+  double X() const;
+  double Y() const;
+  double Z() const;
+  double Pitch() const;
+  double Roll() const;
+  double Yaw() const;
+
+  double& X();
+  double& Y();
+  double& Z();
+  double& Pitch();
+  double& Roll();
+  double& Yaw();
+
+  void X(const double& x);
+  void Y(const double& y);
+  void Z(const double& z);
+  void Pitch(const double& pitch);
+  void Roll(const double& roll);
+  void Yaw(const double& yaw);
+
+private:
+  double _x;
+  double _y;
+  double _z;
+  double _pitch;
+  double _roll;
+  double _yaw;
+
+};
+
+//================================================================
+/*
+* class MengeHandle
+*/
+class MengeHandle
+{
+public:
+
+  MengeHandle(const std::string& resourcePath,
+    const std::string& behaviorFile,
+    const std::string& sceneFile,
+    const float simTimeStep = 0.0
+  )
+  : _resourcePath(resourcePath),
+    _behaviorFile(behaviorFile),
+    _sceneFile(sceneFile),
+    _simTimeStep(simTimeStep),
+    _agentCount(0)
+  {
+
+    _behaviorFile = this->_ResourceFilePath(_behaviorFile);
+    _sceneFile = this->_ResourceFilePath(_sceneFile);
+
+    if (this->_LoadSimulation())
+    {
+      this->initialized = true;
+    }
+    assert(this->initialized);
+  }
+
+  bool initialized = false;
+
+  void SetSimTimeStep(float simTimeStep);
+  float GetSimTimeStep();
+  size_t GetAgentCount();
+  void SimStep(); //proceed one-time simulation step in _sim
+
+  AgentPtr GetAgent(size_t id);
+
+private:
+  std::string _resourcePath;
+  std::string _behaviorFile;
+  std::string _sceneFile;
+  float _simTimeStep;
+  size_t _agentCount;
+  
+  //Have some problem when transfer raw pointer to shared_ptr or unique_ptr
+  std::shared_ptr<Menge::Agents::SimulatorInterface> _sim;
+
+  std::string _ResourceFilePath(const std::string& relativePath) const;
+  bool _LoadSimulation(); //initialize simulatorinterface
+
+};
+
+//================================================================
+/*
+* class ModelTypeDatabase
+*/
+class ModelTypeDatabase
+{
+public:
+
+  ModelTypeDatabase() {}
+
+  struct Record
+  {
+    std::string typeName;
+    std::string fileName;
+    AgentPose3d pose;
+    std::string animation;
+    double animationSpeed;
+    
+    //for ignition
+    std::string modelFilePath;
+  };
+
+  //Create a new record and returns a reference to the record
+  template<typename... Args>
+  Record* Emplace(Args&& ... args){
```

Comment:
I would recommend not using variadic templates for this function, as it is currently only called to create a pair. The calls to this function, seem to be passing in new raw pointers,
https://github.com/FloodShao/traffic_editor/blob/fred/crowd_sim_plugin/building_sim_plugins/building_gazebo_plugins/src/crowd_simulator.cpp#L282
https://github.com/FloodShao/traffic_editor/blob/fred/crowd_sim_plugin/building_sim_plugins/building_ignition_plugins/src/crowd_simulator.cpp#L154

It would be more ideal to let this method handle the creation of the object `Record`, so that you can delete it in the destructor of `ModelTypeDatabase`, currently this object is leaking memory as the allocated memory is not deleted. Do consider using smart pointers to prevent that.

Another way around this is to change the `unordered_map` to hold `<std::string, Record>`, instead of the pointer as value, and `Emplace` returns a mutable reference to the `Record`